RFCs/FS-1145-csharp-collection-expression-support-for-lists-and-sets.md

```diff
@@ -232,7 +236,7 @@ type Set<'T> = …
 
 <!-- Why should we *not* do this? -->
 
-There are no obvious drawbacks to adding this support to FSharp.Core.
+The main arguable drawback of this approach is that the API surface area of FSharp.Core is now technically different depending on the target framework. In practice, we mitigate this by annotating the new APIs with the `CompilerMessageAttribute` and specifying `IsHidden=true`. This means that editors will not suggest `List.Create` and `Set.Create` in autocompletion, and, if a user nonetheless attempts to use them directly, the compiler will emit a warning that these APIs are intended for compiler use only.
```

Comment:
Have you considered the use of the [EditorBrowsableAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.editorbrowsableattribute?view=net-8.0) instead? That's the one I'm most familiar with editors and tooling using as a 'hide me' marker.

Reply:
Yes, that's a possibility. I merely went with `CompilerMessageAttribute` because that's the attribute used for most other constructs defined in FSharp.Core that aren't meant to be used directly outside of it. We could apply both attributes here, although VS and Ionide do already respect `CompilerMessageAttribute` on its own (maybe something in FCS completion is aware of it).

Reply:
I think we would at least want to keep `CompilerMessageAttribute` for the warning (FS1204) either way.

Reply:
I think I'm just going to leave it like this for now.

Only `CompilerMessageAttribute` is used in FSharp.Core at present:

https://github.com/search?q=repo%3Adotnet%2Ffsharp+%5B%3CCompilerMessage+language%3AF%23&type=code

`EditorBrowsableAttribute` is not:

https://github.com/search?q=repo%3Adotnet%2Ffsharp+%5B%3CEditorBrowsable+language%3AF%23&type=code

The name completion logic in FCS accounts for the presence of `CompilerMessageAttribute` with `IsHidden=true` here:

- https://github.com/dotnet/fsharp/blob/df3382ce9d0d66b7e2e3830af5826b0f2dd49289/src/Compiler/Checking/NameResolution.fs#L4336
- https://github.com/dotnet/fsharp/blob/df3382ce9d0d66b7e2e3830af5826b0f2dd49289/src/Compiler/Checking/NameResolution.fs#L4448
- https://github.com/dotnet/fsharp/blob/df3382ce9d0d66b7e2e3830af5826b0f2dd49289/src/Compiler/Checking/AttributeChecking.fs#L452
- https://github.com/dotnet/fsharp/blob/df3382ce9d0d66b7e2e3830af5826b0f2dd49289/src/Compiler/Checking/AttributeChecking.fs#L375
- https://github.com/dotnet/fsharp/blob/df3382ce9d0d66b7e2e3830af5826b0f2dd49289/src/Compiler/Checking/AttributeChecking.fs#L353-L355

If we wanted to, down the line, we could make a separate update to FSharp.Core to add `EditorBrowsableAttribute` to every construct that currently has `CompilerMessageAttribute` with `IsHidden=true`.

Reply:
That's very reasonable - just wanted to raise this as a discussion point because of where I'd seen EditorBrowseable used in other designs on dotnet/designs and dotnet/csharplang.

Reply:
Yeah, and `EditorBrowsableAttribute` would make sense if we also wanted to hide these types and methods from completion in C#, although I'm not sure that's necessary for this feature.

Reply:
And I very much appreciate your taking the time to look and raise the question!

Reply:
> Yeah, and `EditorBrowsableAttribute` would make sense if we also wanted to hide these types and methods from completion in C#, although I'm not sure that's necessary for this feature. The primary reason to hide them in F# is, I think, to maintain parity of the apparent API surface area of FSharp.Core across target frameworks. Do you see a reason to hide them from C#?

For what it's worth, the `Create` methods like this in the BCL are _not_ hidden.